enhancements/multi-arch/heterogeneous-architecture.md

```diff
@@ -0,0 +1,383 @@
+The phases are listed below and details on some components of the phases are described below.
+
+Phase 0: Prototype
+- Aimed at delivering a prototype that users can experiment with.
+- Aimed at Azure/AWS with the primary supported architectures being x86 + ARM.
+- For supporting ARM nodes on Azure, [hyper-v support][BZ-1949614] is required which is landing in RHEL 8.6 and will be included in RHCOS for 4.11.
+- oc adm command changes to support manifestlist payloads.
+- Ability to assemble and publish manifestlist-based release payloads in a reproducible way.
+- Does not include upgrade support.
+- Is not intended for production.
+
+Phase 1: Tech Preview of basic support
+- Release pipeline needs to be setup to construct manifestlist based release payloads.
+- Upgrade support - CVO/Cincinnati support (heterogeneous -> heterogeneous payload upgrades only).
+- Does not include support for:
+  - image registry - internal image registry will not support manifest listed images.
+  - imagestreams importing full manifestlist content (will continue to import a single image when referencing manifestlists).
+
+Phase 2: Full basic functionality
+- Imagestream manifestlist support.
+- OLM support for node selection when deploying an operator and ecosystem improvements.
+- Improve console's content filtering by making it architecture aware and passing that information to OLM.
+- oc debug (and other commands) support for compute architecture.
+- support toolbox.
+- support for Power and Z architectures.
+- Hypershift support for heterogeneous management+worker clusters.
+
+Phase 3: Quality of life improvements
+- First class install support - ability to create non-control-plane-architecture workers at install time and generate machineconfig
+and machineset manifests for the same.
+- Possibly reduce mirroring costs by looking into things like sparse manifestlist.
+- Upgrade support for homogeneous payload -> heterogeneous release payload and vice versa.
+
+### User Stories
+
+- As an enterprise customer looking to save compute resource cost, I want to be able to run part of my application workload on ARM instead of x86.
+- As the owner of an application that runs partially on x86 and partially on ARM, I want to be able to run the entire application within a
+single cluster so I do not need to pay the cost of higher latency between clusters, or worry about the HA-resilience of cross-cluster bridges.
+
+### API Extensions
+
+- Imagestreams API will introduce a new field to indicate that it will reference/import manifestlists. It would also contain a new child manifest
+field which would be for the image corresponding to the control plane architecture.
+- Upgrades to a heterogeneous release payload will require an additional field either in the Infrastructure or the ClusterVersion resource to indicate
+that the user wants to upgrade to a heterogeneous release payload.
```

Comment:
@wking @LalatenduMohanty Can we quickly make a decision here if the CVO needs to know whether the cluster is homogeneous and thus fetches its native architecture update graph or if the CVO will infer from the fact that the release image is a manifestlist image that it's a heterogenous cluster?

I tend to think we should prefer being explicit here, so if for some reason all release images become manifestlist we're not scrambling to address this then.

Reply:
for upgrades, CVO would look at the newly added architecture field in the spec to figure out the type of the cluster. it is not clear how it would know on a fresh install of a cluster with heterogeneous payload - should it look at its own (CVO'S) image and if it is manifestlisted, assume the cluster is heterogeneous?

Reply:
yeah, we not clear how CVO will be know if it is heterogeneous cluster from the pull spec. We will figure that out. But we are moving in that direction of adding an architecture field in ClusterVersion spec.